src/protections/NsfwProtection.ts

```diff
@@ -0,0 +1,85 @@
+/*
+Copyright 2024 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { Protection } from "./IProtection";
+import { Mjolnir } from "../Mjolnir";
+import * as nsfw from 'nsfwjs';
+import {LogLevel} from "matrix-bot-sdk";
+import { node } from '@tensorflow/tfjs-node';
+
+
+export class NsfwProtection extends Protection {
+    settings = {};
+    // @ts-ignore
+    private model: any;
+
+    constructor() {
+        super();
+    }
+
+    async initialize() {
+        this.model = await nsfw.load();
+    }
+
+    public get name(): string {
+        return 'NsfwProtection';
+    }
+
+    public get description(): string {
+        return "Scans all images sent into a protected room to determine if the image is " +
+            "NSFW. If it is, the image will automatically be redacted.";
+    }
+
+    public async handleEvent(mjolnir: Mjolnir, roomId: string, event: any): Promise<any> {
+        if (event['type'] === 'm.room.message') {
+            const content = event['content'] || {};
+            const msgtype = content['msgtype'] || 'm.text';
+            const formattedBody = content['formatted_body'] || '';
+            const isMedia = msgtype === 'm.image' || formattedBody.toLowerCase().includes('<img');
+
+            if (isMedia) {
+                const mxc = content["url"]
```

Comment:
Shouldnt the thumbnail be checked as well? It might be a different image?

Reply:
Possibly, though thumbnails are a bit weird already. This can be revisted in a future PR.

Reply:
if it helps this is how i've laid out Draupnir's attempt to copy this protection, so that it's possible to scrape out more urls from the message https://github.com/the-draupnir-project/Draupnir/blob/gnuxie/nsfwjs/src/protections/NSFWImageProtection.tsx#L55-L71 (incomplete code though, would like to avoid depending on `@tensorflow/tfjs-node` )